repo/fs/refstore.go

```diff
@@ -97,106 +102,92 @@ func (n Refstore) GetRef(get repo.DatasetRef) (repo.DatasetRef, error) {
 }
 
 // DeleteRef removes a name from the store
-func (n Refstore) DeleteRef(del repo.DatasetRef) error {
-	names, err := n.names()
+func (rs Refstore) DeleteRef(del repo.DatasetRef) error {
+	refs, err := rs.refs()
 	if err != nil {
 		return err
 	}
 
-	for i, ref := range names {
+	for i, ref := range refs {
 		if ref.Match(del) {
-			if ref.Path != "" && n.index != nil {
-				if err := n.index.Delete(ref.Path); err != nil {
+			// TODO (b5) - this is search index handling, move this up into base
+			if ref.Path != "" && rs.index != nil {
+				if err := rs.index.Delete(ref.Path); err != nil {
 					log.Debug(err.Error())
 					return err
 				}
 			}
-			names = append(names[:i], names[i+1:]...)
+			refs = append(refs[:i], refs[i+1:]...)
 			break
 		}
 	}
 
-	return n.save(names)
+	return rs.save(refs)
 }
 
 // References gives a set of dataset references from the store
-func (n Refstore) References(limit, offset int) ([]repo.DatasetRef, error) {
-	names, err := n.names()
+func (rs Refstore) References(offset, limit int) ([]repo.DatasetRef, error) {
+	refs, err := rs.refs()
 	if err != nil {
 		return nil, err
 	}
-	res := make([]repo.DatasetRef, limit)
-	for i, ref := range names {
+	res := make(repo.Refs, limit)
+	for i, ref := range refs {
 		if i < offset {
 			continue
 		}
 		if i-offset == limit {
-			return res, nil
+			return []repo.DatasetRef(res), nil
 		}
 		res[i-offset] = ref
 	}
-	return res[:len(names)-offset], nil
+	return res[:len(refs)-offset], nil
 }
 
 // RefCount returns the size of the Refstore
-func (n Refstore) RefCount() (int, error) {
-	names, err := n.names()
+func (rs Refstore) RefCount() (int, error) {
+	// TODO (b5) - there's no need to unmarshal here
+	// could just read the length of the flatbuffer ref vector
+	refs, err := rs.refs()
 	if err != nil {
 		log.Debug(err.Error())
 		return 0, err
 	}
-	return len(names), nil
+	return len(refs), nil
 }
 
-func (n *Refstore) names() ([]repo.DatasetRef, error) {
-	data, err := ioutil.ReadFile(n.filepath(n.file))
+func (rs *Refstore) refs() (repo.Refs, error) {
+	data, err := ioutil.ReadFile(rs.filepath(rs.file))
```

Comment:
We should consider cacheing the deserialized refs, then only reread the ref file is it has a greater mtime than the last time we read it.

Reply:
Great idea. Filed #857